### ISSUE TITLE
Lag compensation

### ISSUE THREAD

**Issue by enneract** [Unvanquished/Unvanquished#423]
Lag compensation is necessary for a multiplayer game, but the implementation inherited from Tremulous has numerous missing features. This issue's purpose is to briefly describe lag compensation and track its progress.
(This issue includes problems mentioned in: #162, #287, #311)
## Lag compensation

A player action (e.g. moving, shooting) is sent to the server, the server processes it (e.g. performs collision detection) and sends back a new snapshot (status of all things in a map). Due to network latency the effects of all player actions are delayed to the point where the game is impossible to play. To counter that, **lag compensation** is required. It consists of two parts:
## Client-side prediction

Since a client is notified instantly about player actions, but a server needs to be notified over a WAN, **clients are ahead of servers** and because waiting for the server to respond is not an option, a client has to **predict** its own future. For example, if a player shoots at a wall, he most likely will hit the wall. Since some things are impossible to predict (e.g. actions of other players) and cgame doesn't have access to all information, **prediction errors** might occur, which have to be corrected (as server is always right).

Implementation progress:
- [ ] **Move all predictable server logic to shared**
- Player state prediction: prediction of...
  - [x] movement (bg_pmove)
  - [x] shooting (bg_pmove)
  - [ ] reloading
  - [ ] commands (buying, selling, evolving)
- Prediction of effects of shooting:
  - [ ] Prediction of missiles
  - [ ] Prediction of hitscan weapons
- [ ] Prediction of menus _(menus shouldn't even be handled by the server anyway)_
- [ ] Prediction of movers
## Server-side unlagging

Performing player actions at the time when the server receives it may result in a different outcome than what the player would have expected. For example, if a player is shooting a moving target and the server is checking for hits a few hundred milliseconds later (so the target will be in a different place), the player might not hit the target even if he's aiming at the dead center. To counter that, the server has to perform the actions at the time when the player caused them, which from server's point of view is **in the past**. This is called **unlagging** (or unlagged).

Unlagging works by remembering the past and going back to it when needed. For example, if a player with 100ms latency shoots, the server looks at the map the way it was 100ms ago and checks for hit then.

Implementation progress:
- Rewinding...
  - [x] players
  - [ ] movers
  - [ ] buildables _(is it even needed?)_
- Unlagging weapons:
  - [x] Unlagging hitscan weapons
  - [ ] Unlagging missiles
- [x] Unlagging player vs player collisions

**NOTE**: Unlagging player vs movers collisions is not needed if cgame can predict movers.

**Comment by bmorel** [comment]
ah, so indeed there is no unlagged mode in unvanquished.

**Comment by DolceTriade** [comment]
This is incorrect. We have unlagged, it just is kinda sketchy at high ping. If we didn't have unlagged, this game would be unplayable at high ping.

**Comment by bmorel** [comment]
I see. Thanks for clarification.

**Comment by DolceTriade** [comment]
175 - 300 ping

Beyond that, it's unplayable.
